### PR TITLE
Updated GPT2 model export script

### PR DESCRIPTION
**Pull request by neginraoof**
# GPT-2

## Use-cases
Transformer-based language model for text generation.

## Description
[GPT-2](https://openai.com/blog/better-language-models/) is a large transformer-based language model with a simple objective: predict the next word, given all of the previous words within some text.

## Model

 |Model        |Download  |Checksum| Download (with sample test data)|ONNX version|Opset version|Accuracy |
|-------------|:--------------|:--------------|:--------------|:--------------|:--------------|:--------------|
|GPT-2       |[522.81 MB](https://github.com/onnx/models/blob/master/text/machine_comprehension/gpt-2/model/model.onnx) | [MD5](https://github.com/onnx/models/blob/master/text/machine_comprehension/gpt-2/model/gpt2-md5.txt)| [438.3 MB](https://github.com/onnx/models/blob/master/text/machine_comprehension/gpt-2/model/GPT-2.tar.gz)| 1.6 | 10 |mAP of [0.024](https://docs.google.com/spreadsheets/d/1sryqufw2D0XlUH4sq3e9Wnxu5EAQkaohzrJbd5HdQ_w/edit#gid=0)|

## Inference
The script for ONNX model conversion and ONNXRuntime inference is [here](https://github.com/onnx/models/blob/master/text/machine_comprehension/gpt-2/GPT2-export.py)

### Input to model
Sequence of words as a string. Example: "Here is some text to encode : Hello World", tokenized by Byte-Pair-Encoding.
**input_ids**: Indices of input tokens in the vocabulary. It's a long tensor of dynamic shape (batch_size, sequence_length).

### Preprocessing steps
Use ```tokenizer.encode()``` to encode the input text:
```python
text = "Here is some text to encode : Hello World"
tokenizer = GPT2Tokenizer.from_pretrained('gpt2')
tokens_tensor = torch.tensor([torch.tensor(tokenizer.encode(text))])
```

### Output of model
**last_hidden_state**: Sequence of hidden-states at the last layer of the model. It's a float tensor of size (batch_size, sequence_length, hidden_size).
**past**: pre-computed hidden-states. It's a list of tensors (key and values in the attention blocks) of size (batch_size, num_heads, sequence_length, sequence_length), one per each layer.

Output of this model is the tuple (last_hidden_state, past)

Note that output_hidden_states=False and output_attentions=False in the PretrainedConfig configs.

### Postprocessing steps
```python
outputs = model(input_ids)
last_hidden_states = outputs[0]
```
<hr>

## Dataset (Train and validation)
The original model from OpenAI is pretrained on a dataset of [8 million web pages](https://openai.com/blog/better-language-models).
The pretrained model is referenced in  [huggingface/transformers](https://github.com/huggingface/transformers/blob/master/transformers/modeling_gpt2.py) repository as a causal (unidirectional) transformer pre-trained using language modeling on a very large corpus of ~40 GB of text data.
https://s3.amazonaws.com/models.huggingface.co/bert/gpt2-pytorch_model.bin

<hr>

## Validation accuracy
Metric and benchmarking details are provided by HuggingFace in this [post](https://medium.com/huggingface/benchmarking-transformers-pytorch-and-tensorflow-e2917fb891c2).
<hr>

## Publication/Attribution
Alec Radford, Jeffrey Wu, Rewon Child, David Luan, Dario Amodei, andIlya Sutskever. Language Models are Unsupervised Multitask Learners. 2019.

## References
This model is converted directly from [huggingface/transformers](https://github.com/huggingface/transformers/blob/master/src/transformers/modeling_gpt2.py)
<hr>

## Contributors
Negin Raoof
<hr>

## License
Apache 2.0 License
<hr>

### CODE REVIEW

**Comment by neginraoof** [comment]
cc @EmmaNingMS @prasanthpul 
I needed to update the script that we use for exporting GPT2 model, for it to save input/output tensorproto files that are compatible with onnxruntime. Could you please help review this PR?

cc @spandantiwari
Thanks,
Negin

File: text/machine_comprehension/gpt-2/GPT2-export.py [hunk 1]
```diff
@@ -13,6 +14,22 @@
 data_dir = 'test_data_set_0'
 
 
+def f(t):
```
Comment (spandantiwari) [inline text/machine_comprehension/gpt-2/GPT2-export.py]: I suggest we provide more descriptive names for these methods.

File: text/machine_comprehension/gpt-2/GPT2-export.py [hunk 2]
```diff
@@ -1,5 +1,6 @@
 import torch
```
Comment (spandantiwari) [inline text/machine_comprehension/gpt-2/GPT2-export.py]: Can we make the updates to the script in ShuffleNet_v2 model as well?
https://github.com/onnx/models/blob/master/vision/classification/shufflenet_v2/shufflenet-export.py

File: text/machine_comprehension/gpt-2/GPT2-export.py [hunk 3]
```diff
@@ -21,15 +38,15 @@ def save_tensor_proto(file_path, name, data):
...
     if isinstance(data_list, torch.autograd.Variable) or isinstance(data_list, torch.Tensor):
         data_list = [data_list]
     for i, d in enumerate(data_list):
         d = d.data.cpu().numpy()
-        save_tensor_proto(os.path.join(test_data_dir, '{0}_{1}.pb'.format(prefix, i)), prefix + str(i+1), d)
+        save_tensor_proto(os.path.join(test_data_dir, '{0}_{1}.pb'.format(prefix, i)), names[i], d)
 
 
-def save(name, model, inputs, outputs, input_names=['input'], output_names=['output'], **kwargs):
+def save(name, model, inputs, outputs, input_names=None, output_names=None, **kwargs):
```
Comment (vinitra) [inline text/machine_comprehension/gpt-2/GPT2-export.py]: more detailed function name might be helpful

File: text/machine_comprehension/gpt-2/GPT2-export.py (hunk 1 above)
Comment (vinitra) [reply text/machine_comprehension/gpt-2/GPT2-export.py]: Seconded. Arguments, also.

File: text/machine_comprehension/gpt-2/GPT2-export.py [hunk 4]
```diff
@@ -93,23 +119,47 @@ def gpt2_test():
...
+            f_ = os.path.join(data_dir, '{0}_{1}.pb'.format("output", i))
+            tensor = onnx.TensorProto()
+            with open(f_, 'rb') as file:
+                tensor.ParseFromString(file.read())
+            outputs.append(numpy_helper.to_array(tensor))
+
+        inference(model_dir, inputs, outputs)
+
+        # Test exported model with a new input
+        print("****** Feeding model with new input *******")
```
Comment (vinitra) [inline text/machine_comprehension/gpt-2/GPT2-export.py]: nit: consistent comment style format: ==== Text ==== instead of ****

File: text/machine_comprehension/gpt-2/README.md [hunk 5]
```diff
@@ -61,7 +61,7 @@ Metric and benchmarking details are provided by HuggingFace in this [post](https
 Alec Radford, Jeffrey Wu, Rewon Child, David Luan, Dario Amodei, andIlya Sutskever. Language Models are Unsupervised Multitask Learners. 2019.
 
 ## References
-This model is converted directly from [huggingface/transformers](https://github.com/huggingface/transformers/blob/master/transformers/modeling_gpt2.py)
+This model is converted directly from [huggingface/transformers](https://github.com/huggingface/transformers/blob/master/src/transformers/modeling_gpt2.py)
```
Comment (vinitra) [inline text/machine_comprehension/gpt-2/README.md]: Period at end of sentence

File: text/machine_comprehension/gpt-2/GPT2-export.py (hunk 2 above)
Comment (neginraoof) [reply text/machine_comprehension/gpt-2/GPT2-export.py]: Sure. Done.

**Comment by neginraoof** [comment]
Thanks @vinitra
